[SPARK-31238][SQL] Rebase dates to/from Julian calendar in write/read for ORC datasource

MaxGekk opened the pull request:
### What changes were proposed in this pull request?

This PR (SPARK-31238) aims the followings.
1. Modified ORC Vectorized Reader, in particular, OrcColumnVector v1.2 and v2.3. After the changes, it uses `DateTimeUtils. rebaseJulianToGregorianDays()` added by https://github.com/apache/spark/pull/27915 . The method performs rebasing days from the hybrid calendar (Julian + Gregorian) to Proleptic Gregorian calendar. It builds a local date in the original calendar, extracts date fields `year`, `month` and `day` from the local date, and builds another local date in the target calendar. After that, it calculates days from the epoch `1970-01-01` for the resulted local date.
2. Introduced rebasing dates while saving ORC files, in particular, I modified `OrcShimUtils. getDateWritable` v1.2 and v2.3, and returned `DaysWritable` instead of Hive's `DateWritable`. The `DaysWritable` class was added by the PR https://github.com/apache/spark/pull/27890 (and fixed by https://github.com/apache/spark/pull/27962). I moved `DaysWritable` from `sql/hive` to `sql/core` to re-use it in ORC datasource.

### Why are the changes needed?
For the backward compatibility with Spark 2.4 and earlier versions. The changes allow users to read dates/timestamps saved by previous version, and get the same result.

### Does this PR introduce any user-facing change?
Yes. Before the changes, loading the date `1200-01-01` saved by Spark 2.4.5 returns the following:
```scala
scala> spark.read.orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc").show(false)
+----------+
|dt        |
+----------+
|1200-01-08|
+----------+
```
After the changes
```scala
scala> spark.read.orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc").show(false)
+----------+
|dt        |
+----------+
|1200-01-01|
+----------+
```

### How was this patch tested?
- By running `OrcSourceSuite` and `HiveOrcSourceSuite`.
- Add new test `SPARK-31238: compatibility with Spark 2.4 in reading dates` to `OrcSuite` which reads an ORC file saved by Spark 2.4.5 via the commands:
```shell
$ export TZ="America/Los_Angeles"
```
```scala
scala> sql("select cast('1200-01-01' as date) dt").write.mode("overwrite").orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc")
scala> spark.read.orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc").show(false)
+----------+
|dt        |
+----------+
|1200-01-01|
+----------+
```
- Add round trip test `SPARK-31238: rebasing dates in write`. The test `SPARK-31238: compatibility with Spark 2.4 in reading dates` confirms rebasing in read. So, we can check rebasing in write.

MaxGekk committed:
Add a test

MaxGekk committed:
Fix v2.3 OrcColumnVector

MaxGekk committed:
Fix v1.2 OrcColumnVector

SparkQA commented:
**[Test build #120324 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120324/testReport)** for PR 28016 at commit [`ecef05d`](https://github.com/apache/spark/commit/ecef05d2f02b7df5ac653b743209f87ae7a68841).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25035/
Test PASSed.

SparkQA commented:
**[Test build #120324 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120324/testReport)** for PR 28016 at commit [`ecef05d`](https://github.com/apache/spark/commit/ecef05d2f02b7df5ac653b743209f87ae7a68841).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120324/
Test FAILed.

MaxGekk committed:
Rebase in write

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25057/
Test PASSed.

SparkQA commented:
**[Test build #120345 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120345/testReport)** for PR 28016 at commit [`47d8588`](https://github.com/apache/spark/commit/47d8588510d453040d316abe560061f2e4242ea6).

dongjoon-hyun commented:
So, does this happen only in vectorized reader, @MaxGekk ?

SparkQA commented:
**[Test build #120345 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120345/testReport)** for PR 28016 at commit [`47d8588`](https://github.com/apache/spark/commit/47d8588510d453040d316abe560061f2e4242ea6).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120345/
Test FAILed.

MaxGekk commented:
> So, does this happen only in vectorized reader

@dongjoon-hyun Correct, the regular reader uses `DateTimeUtils.fromJavaDate`:
https://github.com/apache/spark/blob/300ec1a74cb14867c22e616e657566d510426331/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala#L111

And the PRs #27807 and #27980 introduced date rebasing in `fromJavaDate()`.

Also, I fixed ORC writer, and added a round trip test for both vectorized and non-vectorized readers.

MaxGekk commented:
I think the failure doesn't related to my changes: https://github.com/apache/spark/pull/28016#issuecomment-603958162 . I will update the PR desciption shortly, so the PR will be ready for review. /cc @cloud-fan

MaxGekk commented:
jenkins, retest this, please

SparkQA commented:
**[Test build #120367 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120367/testReport)** for PR 28016 at commit [`47d8588`](https://github.com/apache/spark/commit/47d8588510d453040d316abe560061f2e4242ea6).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25076/
Test PASSed.

MaxGekk commented:
@cloud-fan @dongjoon-hyun @bersprockets Please, review this PR.

MaxGekk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala:
```diff
@@ -35,11 +35,12 @@ import org.apache.spark.sql.catalyst.util.DateTimeUtils.{rebaseGregorianToJulian
     var julianDays: Int)
   extends DateWritable {
 
+  def this() = this(0, 0)
```
I assume that `gregorianDays` and `julianDays` will be set later via the `set` method.

SparkQA commented:
**[Test build #120367 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120367/testReport)** for PR 28016 at commit [`47d8588`](https://github.com/apache/spark/commit/47d8588510d453040d316abe560061f2e4242ea6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120367/
Test PASSed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #120379 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120379/testReport)** for PR 28016 at commit [`47d8588`](https://github.com/apache/spark/commit/47d8588510d453040d316abe560061f2e4242ea6).

dongjoon-hyun reviewed (dismissed):
+1, LGTM (Pending Jenkins with `hive1.2` to make it sure.)
Thank you for swift fix, @MaxGekk !

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25089/
Test PASSed.

SparkQA commented:
**[Test build #120379 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120379/testReport)** for PR 28016 at commit [`47d8588`](https://github.com/apache/spark/commit/47d8588510d453040d316abe560061f2e4242ea6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120379/
Test FAILed.

dongjoon-hyun commented:
Hi, @MaxGekk . 
Could you fix the compilation failures in `hive-1.2` profile?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala:
```diff
@@ -35,11 +35,12 @@ import org.apache.spark.sql.catalyst.util.DateTimeUtils.{rebaseGregorianToJulian
+class DaysWritable(
     var gregorianDays: Int,
     var julianDays: Int)
   extends DateWritable {
```
This causes a failure in `hive-1.2` profile due to the following.
- `org.apache.hadoop.hive.serde2.io.DateWritable`
- `org.apache.orc.storage.serde2.io.DateWritable`
```
/home/jenkins/workspace/SparkPullRequestBuilder@3/sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcShimUtils.scala:54: type mismatch;
[error]  found   : org.apache.spark.sql.execution.datasources.DaysWritable
[error]  required: org.apache.orc.storage.serde2.io.DateWritable
[error]         result
[error]         ^
[info] org.apache.spark.sql.execution.datasources.DaysWritable <: org.apache.orc.storage.serde2.io.DateWritable?
[info] false
[error] /home/jenkins/workspace/SparkPullRequestBuilder@3/sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcShimUtils.scala:57: type mismatch;
[error]  found   : org.apache.spark.sql.execution.datasources.DaysWritable
[error]  required: org.apache.orc.storage.serde2.io.DateWritable
[error]         new DaysWritable(getter.getInt(ordinal))
```

cloud-fan replied:
We probably need to duplicate this class in v1.2 and v2.3 source code.

MaxGekk replied:
I cloned `DaysWritable` for Hive ORC v1.2, and added a comment to the cloned class.

cloud-fan commented on sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java:
```diff
@@ -130,7 +138,13 @@ public short getShort(int rowId) {
   public int getInt(int rowId) {
-    return (int) longData.vector[getRowIndex(rowId)];
+    int index = getRowIndex(rowId);
+    int value = (int) longData.vector[index];
```
nit: `int value = (int) longData.vector[getRowIndex(rowId)];`

cloud-fan commented on sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java:
```diff
@@ -42,6 +43,7 @@
   private DecimalColumnVector decimalData;
   private TimestampColumnVector timestampData;
```
does ORC have `DateColumnVector`?

MaxGekk replied:
Yes, it does, but:
1. `DateColumnVector` doesn't have a method similar to `asScratchTimestamp` in `TimestampColumnVector`
2. We don't need to build `java.sql.Date` from serialized days to perform rebasing. It is unnecessary overhead.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala:
```diff
@@ -32,6 +32,7 @@ import org.apache.spark.sql.AnalysisException
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.datasources.DaysWritable
```
unnecessary change

MaxGekk replied:
If I remove it, the build fails with:
```
Error:(1012, 44) not found: type DaysWritable
  private def getDateWritable(value: Any): DaysWritable =
```
I moved `DaysWritable` from `sql/hive` to `sql/core` to reuse it in ORC.

MaxGekk committed:
Copy DaysWritable to v1.2

MaxGekk committed:
Remove `index` variable in OrcColumnVector

SparkQA commented:
**[Test build #120403 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120403/testReport)** for PR 28016 at commit [`3b1b791`](https://github.com/apache/spark/commit/3b1b791da17cc256c53310ad9f0f58d15da7689a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25112/
Test PASSed.

SparkQA commented:
**[Test build #120403 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120403/testReport)** for PR 28016 at commit [`3b1b791`](https://github.com/apache/spark/commit/3b1b791da17cc256c53310ad9f0f58d15da7689a).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120403/
Test FAILed.

MaxGekk commented:
jenkins, retest this, please

SparkQA commented:
**[Test build #120408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120408/testReport)** for PR 28016 at commit [`3b1b791`](https://github.com/apache/spark/commit/3b1b791da17cc256c53310ad9f0f58d15da7689a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25116/
Test PASSed.

MaxGekk committed:
Added a comment to `DaysWritable` v1.2

SparkQA commented:
**[Test build #120416 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120416/testReport)** for PR 28016 at commit [`c8a897a`](https://github.com/apache/spark/commit/c8a897abd4a0fae72d42335951ff77b1f054c655).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25124/
Test PASSed.

SparkQA commented:
**[Test build #120408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120408/testReport)** for PR 28016 at commit [`3b1b791`](https://github.com/apache/spark/commit/3b1b791da17cc256c53310ad9f0f58d15da7689a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120408/
Test PASSed.

SparkQA commented:
**[Test build #120416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120416/testReport)** for PR 28016 at commit [`c8a897a`](https://github.com/apache/spark/commit/c8a897abd4a0fae72d42335951ff77b1f054c655).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120416/
Test PASSed.

dongjoon-hyun reviewed (commented):
Please move `sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala` to `sql/core/v2.3/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala`.

dongjoon-hyun commented on sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala:
```diff
@@ -0,0 +1,79 @@
+ * This is a clone of `org.apache.spark.sql.execution.datasources.DaysWritable`.
+ * The class is cloned because Hive ORC v1.2 uses different `DateWritable`:
+ *   - v1.2: `org.apache.orc.storage.serde2.io.DateWritable`
+ *   - v2.3 and `HiveInspectors`: `org.apache.hadoop.hive.serde2.io.DateWritable`
```
We don't need the above 4 line comments because `v1.2` and `v2.3` folder structure already is designed for that.

MaxGekk replied:
If I move it to 2.3, and enable hive-1.2, `HiveInspectors` will use `DaysWritable` from v1.2 which is wrong or I miss something in your proposal?

dongjoon-hyun replied:
Got it. Let me check again.

dongjoon-hyun replied:
? Sorry, why do you move this to 2.3?

dongjoon-hyun replied:
What I ask is keep this AS-IS, and move the other `DaysWritable`.

dongjoon-hyun replied:
I'm checking again from my side.

MaxGekk replied:
This is what I tried https://github.com/MaxGekk/spark/pull/26, and it fails when I enable hive-1.2:
```
build/sbt -Phive-1.2 clean package
```
Feel free to open a PR for this PR if you mean something different.

dongjoon-hyun commented on sql/core/v2.3/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcShimUtils.scala:
```diff
@@ -26,6 +26,7 @@ import org.apache.hadoop.hive.ql.io.sarg.PredicateLeaf.{Operator => OrcOperator}
 import org.apache.hadoop.hive.serde2.io.{DateWritable, HiveDecimalWritable}
 
 import org.apache.spark.sql.catalyst.expressions.SpecializedGetters
+import org.apache.spark.sql.execution.datasources.DaysWritable
```
After moving `DaysWritable.scala` to here. I guess we don't need this line. Please try to remove.

MaxGekk commented:
> Please move ... to v2.3

Why? `DaysWritable` is used in sql/hive, in `HiveInspectors`, just in case.

dongjoon-hyun commented:
Ur, are you asking why we keep `v1.2` and `v2.3` structure?

dongjoon-hyun commented:
My bad. I overlooked the difference in `HiveInspectors` in `hive-1.2` profile.

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @MaxGekk and @cloud-fan .
Merged to master/3.0.

dongjoon-hyun commented:
Hi, @MaxGekk . 
Unfortunately, this seems to break all Maven build in master/branch-3.0.
Could you take a look at this?

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcTest.scala:
```diff
@@ -133,4 +133,9 @@ abstract class OrcTest extends QueryTest with FileBasedDataSourceTest with Befor
   }
+
+  protected def readResourceOrcFile(name: String): DataFrame = {
+    val url = Thread.currentThread().getContextClassLoader.getResource(name)
```
This might be the root cause.

dongjoon-hyun commented:
I made a follow-up.
- https://github.com/apache/spark/pull/28059
